File: StocksAppSolution/StocksApp.Core/Services/FinnhubService.cs
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;
using StocksApp.Core.ServiceContracts;

namespace StocksApp.Core.Services
{
    public class FinnhubService : IFinnHubService
    {
        private readonly IConfiguration _configuration;
        private readonly HttpClient _httpClient;
        private readonly string _apiKey;

        public FinnhubService(IConfiguration configuration, HttpClient httpClient)
        {
            _configuration = configuration;
            _httpClient = httpClient;
            _apiKey = _configuration["FinnhubApiKey"];
        }
        private string BuildUrl(string endpoint, string stockSymbol)
        {
            return $"https://finnhub.io/api/v1/{endpoint}?symbol={stockSymbol}&token={_apiKey}";
        }
        private async Task<Dictionary<string,object>?> GetData(string url)
        {
            var response = await _httpClient.GetAsync(url);

            if (!response.IsSuccessStatusCode)
            {
                throw new Exception($"Failed to fetch company profile.\n Status code: {response.StatusCode}");
            }
            try
            {
                string json = await response.Content.ReadAsStringAsync();
                Dictionary<string, object>? companyProfile = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
                return companyProfile;
            }
            catch(Exception ex)
            {
                throw new Exception($"Error deserializing company profile: {ex.Message}");
            }
        }
        public async Task<Dictionary<string, object>?> GetCompanyProfile(string stockSymbol) 
        {
            if(string.IsNullOrEmpty(stockSymbol))
            {
                throw new ArgumentException("Stock symbol cannot be empty or null", nameof(stockSymbol));
            }
            
            string url = BuildUrl("stock/profile2", stockSymbol);

            var companyData = await GetData(url);

            return companyData;
        }

        public async Task<Dictionary<string, object>?> GetStockPriceQuote(string stockSymbol)
        {
            if (string.IsNullOrEmpty(stockSymbol))
            {
                throw new ArgumentException("Stock symbol cannot be empty or null", nameof(stockSymbol));
            }
            string url = BuildUrl("quote", stockSymbol);

            var companyData = await GetData(url);
            return companyData;
        }
    }
}

File: StocksAppSolution/StocksApp.WebApi/ReactWebClient/src/app/services/finnhubService.ts
import { StockProfile, StockQuote } from "../types/trading";

// Use the provided API token or replace with your own from https://finnhub.io/dashboard
const FINNHUB_TOKEN = "cc676uaad3i9rj8tb1s0";

export const finnhubService = {
  async getCompanyProfile(stockSymbol: string): Promise<StockProfile | null> {
    try {
      const response = await fetch(
        `https://finnhub.io/api/v1/stock/profile2?symbol=${stockSymbol}&token=${FINNHUB_TOKEN}`
      );
      if (!response.ok) {
        throw new Error("Failed to fetch company profile");
      }
      const data = await response.json();
      return data;
    } catch (error) {
      console.error("Error fetching company profile:", error);
      return null;
    }
  },

  async getStockPriceQuote(stockSymbol: string): Promise<StockQuote | null> {
    try {
      const response = await fetch(
        `https://finnhub.io/api/v1/quote?symbol=${stockSymbol}&token=${FINNHUB_TOKEN}`
      );
      if (!response.ok) {
        throw new Error("Failed to fetch stock quote");
      }
      const data = await response.json();
      return data;
    } catch (error) {
      console.error("Error fetching stock quote:", error);
      return null;
    }
  },

  getWebSocketUrl(): string {
    return `wss://ws.finnhub.io?token=${FINNHUB_TOKEN}`;
  },
};
